File: frontend/src/app/record/record.service.ts
import { Injectable } from '@angular/core';
import { HttpClient } from '@angular/common/http';
import { Observable } from 'rxjs';

export interface CreateRecordRequest {
  batchId: number;
  tempMass: number;
  tempAmbient: number;
  humidityMass: number;
  humidityAmbient: number;
  phLevel: number;
  isTurned: boolean;
  operatorObservations: string;
}

export interface RecordResponse {
  id: number;
  batchId: number;
  recordedAt: string;
  tempMass: number;
  tempAmbient: number;
  humidityMass: number;
  humidityAmbient: number;
  phLevel: number;
  isTurned: boolean;
  operatorObservations: string;
}

@Injectable({
  providedIn: 'root'
})
export class RecordService {
  private apiUrl = '/api/records';

  constructor(private http: HttpClient) { }

  createRecord(record: CreateRecordRequest): Observable<RecordResponse> {
    return this.http.post<RecordResponse>(this.apiUrl, record);
  }

  getRecordsByBatch(batchId: number): Observable<RecordResponse[]> {
    return this.http.get<RecordResponse[]>(`${this.apiUrl}/batch/${batchId}`);
  }
}

File: backend/src/main/java/com/cloudferm/service/RecordService.java
package com.cloudferm.service;

import com.cloudferm.dto.CreateRecordRequest;
import com.cloudferm.dto.RecordResponse;
import com.cloudferm.model.Batch;
import com.cloudferm.model.DailyRecord;
import com.cloudferm.repository.BatchRepository;
import com.cloudferm.repository.RecordRepository;
import lombok.RequiredArgsConstructor;
import org.springframework.stereotype.Service;

import java.time.LocalDateTime;
import java.util.List;
import java.util.stream.Collectors;

@Service
@RequiredArgsConstructor
public class RecordService {

    private final RecordRepository recordRepository;
    private final BatchRepository batchRepository;

    public RecordResponse createRecord(CreateRecordRequest request) {
        Batch batch = batchRepository.findById(request.getBatchId())
                .orElseThrow(() -> new RuntimeException("Batch not found"));

        DailyRecord record = DailyRecord.builder()
                .batch(batch)
                .recordedAt(LocalDateTime.now())
                .tempMass(request.getTempMass())
                .tempAmbient(request.getTempAmbient())
                .humidityMass(request.getHumidityMass())
                .humidityAmbient(request.getHumidityAmbient())
                .phLevel(request.getPhLevel())
                .isTurned(request.isTurned())
                .operatorObservations(request.getOperatorObservations())
                .build();

        DailyRecord savedRecord = recordRepository.save(record);
        return mapToResponse(savedRecord);
    }

    public List<RecordResponse> getRecordsByBatch(Long batchId) {
        return recordRepository.findByBatchIdOrderByRecordedAtAsc(batchId).stream()
                .map(this::mapToResponse)
                .collect(Collectors.toList());
    }

    private RecordResponse mapToResponse(DailyRecord record) {
        return RecordResponse.builder()
                .id(record.getId())
                .batchId(record.getBatch().getId())
                .recordedAt(record.getRecordedAt())
                .tempMass(record.getTempMass())
                .tempAmbient(record.getTempAmbient())
                .humidityMass(record.getHumidityMass())
                .humidityAmbient(record.getHumidityAmbient())
                .phLevel(record.getPhLevel())
                .isTurned(record.isTurned())
                .operatorObservations(record.getOperatorObservations())
                .build();
    }
}
